File: db4ounit.extensions/src/db4ounit/extensions/fixtures/AbstractSoloDb4oFixture.java
/* Copyright (C) 2004 - 2006 Versant Inc. http://www.db4o.com */

package db4ounit.extensions.fixtures;

import com.db4o.*;
import com.db4o.config.*;
import com.db4o.ext.*;
import com.db4o.internal.*;
import com.db4o.internal.threading.*;

import db4ounit.*;
import db4ounit.extensions.*;

public abstract class AbstractSoloDb4oFixture extends AbstractDb4oFixture {

	private ExtObjectContainer _db;
	
	protected AbstractSoloDb4oFixture() {
	}
	
	public final void open(Db4oTestCase testInstance) {
		Assert.isNull(_db);
		final Configuration config = cloneConfiguration();
		applyFixtureConfiguration(testInstance, config);
		_db=createDatabase(config).ext();
		listenToUncaughtExceptions(threadPool());
		
		postOpen(testInstance);
	}

	private ThreadPool4 threadPool() {
	    return threadPoolFor(_db);
    }

	public void close() throws Exception {
		try {
			preClose();
		}
		finally {
		
		if (null != _db) {
				Assert.isTrue(_db.close());
				try {
					threadPool().join(3000);
				} finally {
					_db = null;
				}
			}
		}
	}	

	public boolean accept(Class clazz) {
		return !OptOutSolo.class.isAssignableFrom(clazz);
	}

	public ExtObjectContainer db() {
		return _db;
	}
	
	public LocalObjectContainer fileSession() {
		return (LocalObjectContainer)_db;
	}
	
	public void configureAtRuntime(RuntimeConfigureAction action) {
		action.apply(config());
	}
	
	protected void preClose() {
	}
	
	protected void postOpen(Db4oTestCase testInstance) {
	}
	
	protected abstract ObjectContainer createDatabase(Configuration config);
	

}

File: db4o.net/Db4oUnit.Extensions/Db4oUnit.Extensions/Fixtures/AbstractSoloDb4oFixture.cs
/* Copyright (C) 2004 - 2011  Versant Inc.  http://www.db4o.com */

using System;
using Db4oUnit;
using Db4oUnit.Extensions;
using Db4oUnit.Extensions.Fixtures;
using Db4objects.Db4o;
using Db4objects.Db4o.Config;
using Db4objects.Db4o.Ext;
using Db4objects.Db4o.Internal;
using Db4objects.Db4o.Internal.Threading;

namespace Db4oUnit.Extensions.Fixtures
{
	public abstract class AbstractSoloDb4oFixture : AbstractDb4oFixture
	{
		private IExtObjectContainer _db;

		protected AbstractSoloDb4oFixture()
		{
		}

		public sealed override void Open(IDb4oTestCase testInstance)
		{
			Assert.IsNull(_db);
			IConfiguration config = CloneConfiguration();
			ApplyFixtureConfiguration(testInstance, config);
			_db = CreateDatabase(config).Ext();
			ListenToUncaughtExceptions(ThreadPool());
			PostOpen(testInstance);
		}

		private IThreadPool4 ThreadPool()
		{
			return ThreadPoolFor(_db);
		}

		/// <exception cref="System.Exception"></exception>
		public override void Close()
		{
			try
			{
				PreClose();
			}
			finally
			{
				if (null != _db)
				{
					Assert.IsTrue(_db.Close());
					try
					{
						ThreadPool().Join(3000);
					}
					finally
					{
						_db = null;
					}
				}
			}
		}

		public override bool Accept(Type clazz)
		{
			return !typeof(IOptOutSolo).IsAssignableFrom(clazz);
		}

		public override IExtObjectContainer Db()
		{
			return _db;
		}

		public override LocalObjectContainer FileSession()
		{
			return (LocalObjectContainer)_db;
		}

		public override void ConfigureAtRuntime(IRuntimeConfigureAction action)
		{
			action.Apply(Config());
		}

		protected virtual void PreClose()
		{
		}

		protected virtual void PostOpen(IDb4oTestCase testInstance)
		{
		}

		protected abstract IObjectContainer CreateDatabase(IConfiguration config);
	}
}
